mmcv/runner/hooks/optimizer.py

```diff
@@ -34,128 +42,234 @@ def after_train_iter(self, runner):
-        if loss_scale == 'dynamic':
-            self.loss_scaler = LossScaler(mode='dynamic')
-        elif isinstance(loss_scale, float):
-            self.loss_scaler = LossScaler(init_scale=loss_scale, mode='static')
-        elif isinstance(loss_scale, dict):
-            self.loss_scaler = LossScaler(**loss_scale)
-        else:
-            raise ValueError('loss_scale must be of type float, dict, or '
-                             f'"dynamic", got {loss_scale}')
-
-    def before_run(self, runner):
-        """Preparing steps before Mixed Precision Training.
-
-        1. Make a master copy of fp32 weights for optimization.
-        2. Convert the main model from fp32 to fp16.
-        """
-        # keep a copy of fp32 weights
-        old_groups = runner.optimizer.param_groups
-        runner.optimizer.param_groups = copy.deepcopy(
-            runner.optimizer.param_groups)
-        state = defaultdict(dict)
-        p_map = {
-            old_p: p
-            for old_p, p in zip(
-                chain(*(g['params'] for g in old_groups)),
-                chain(*(g['params'] for g in runner.optimizer.param_groups)))
-        }
-        for k, v in runner.optimizer.state.items():
-            state[p_map[k]] = v
-        runner.optimizer.state = state
-        # convert model to fp16
-        wrap_fp16_model(runner.model)
-
-    def copy_grads_to_fp32(self, fp16_net, fp32_weights):
-        """Copy gradients from fp16 model to fp32 weight copy."""
-        for fp32_param, fp16_param in zip(fp32_weights, fp16_net.parameters()):
-            if fp16_param.grad is not None:
-                if fp32_param.grad is None:
-                    fp32_param.grad = fp32_param.data.new(fp32_param.size())
-                fp32_param.grad.copy_(fp16_param.grad)
-
-    def copy_params_to_fp16(self, fp16_net, fp32_weights):
-        """Copy updated params from fp32 weight copy to fp16 model."""
-        for fp16_param, fp32_param in zip(fp16_net.parameters(), fp32_weights):
-            fp16_param.data.copy_(fp32_param.data)
+if TORCH_VERSION != 'parrots' and TORCH_VERSION >= '1.6.0':
 
-    def after_train_iter(self, runner):
-        """Backward optimization steps for Mixed Precision Training. For
-        dynamic loss scaling, please refer `loss_scalar.py`
-
-        1. Scale the loss by a scale factor.
-        2. Backward the loss to obtain the gradients (fp16).
-        3. Copy gradients from the model to the fp32 weight copy.
-        4. Scale the gradients back and update the fp32 weight copy.
-        5. Copy back the params from fp32 weight copy to the fp16 model.
+    @HOOKS.register_module()
+    class Fp16OptimizerHook(OptimizerHook):
+        """FP16 optimizer hook (using PyTorch's implementation).
+
+        If you are using PyTorch >= 1.6, torch.cuda.amp is used as the backend,
+        to take care of the optimization procedure.
+
+        Args:
+            loss_scale (float | str | dict): Scale factor multiplied with loss.
```

Comment:
The docstring is not consistent with the real arguments.

Reply:
fixed